### PR TITLE
Back button fix

### PR DESCRIPTION
**Pull request by oliwiaKot**
This is the feature branch for VEOSS-646 (back button fix)
Sections to complete:

- [x] corrections
- [x] sales from NI
- [x] sales from EU
- [x] final CYA

Before merging please review the functionality with the design team. Outstanding questions:

- [ ] where should we redirect when vat rates are missing
- [ ] where should we redirect if net value of sales/vat on sales is missing
- [ ] what should be the content of the "CheckSalesFromEu/Ni" page when there is missing information on sales - do we show the table or just the prompt, what should the prompt say
- [ ] make sure NI and EU journeys are consistent

### CODE REVIEW

File: app/connectors/VatReturnConnector.scala [hunk 1]
```diff
@@ -38,7 +38,7 @@ class VatReturnConnector @Inject()(config: Configuration, httpClient: HttpClient
     httpClient.POST[VatReturnRequest, VatReturnResponse](url, vatReturnRequest)
   }
 
-  def submitWithCorrection(vatReturnRequest: VatReturnWithCorrectionRequest)(implicit hc: HeaderCarrier): Future[VatReturnWithCorrectionResponse] = {
+  def submit(vatReturnRequest: VatReturnWithCorrectionRequest)(implicit hc: HeaderCarrier): Future[VatReturnWithCorrectionResponse] = {
```
Comment (darbyo) [inline app/connectors/VatReturnConnector.scala:41]: Is there a particular reason for renaming this?

File: app/controllers/CheckYourAnswersController.scala [hunk 2]
```diff
@@ -151,81 +156,143 @@ class CheckYourAnswersController @Inject()(
...
-            val errorMessages = errorList.map(_.errorMessage).mkString("\n")
-            logger.error(s"Unable to create a VAT return request from user answers: $errorMessages")
-
-            Redirect(routes.JourneyRecoveryController.onPageLoad()).toFuture
-        }
+
+      val redirectToFirstError = getRedirect(validate(period), period).headOption
+
+      (redirectToFirstError, incompletePromptShown) match {
+        case (Some(c), true) => Future.successful(Redirect(c))
```
Comment (darbyo) [inline app/controllers/CheckYourAnswersController.scala:184]: could we give c a more meaningful name

File: app/controllers/CheckYourAnswersController.scala [hunk 3]
```diff
@@ -151,81 +156,143 @@ class CheckYourAnswersController @Inject()(
...
+            case (_, Some(Invalid(errors))) =>
+              val errorList = errors.toChain.toList
+              val errorMessages = errorList.map(_.errorMessage).mkString("\n")
+              logger.error(s"Unable to create a Corrections request from user answers: $errorMessages")
+              Redirect(routes.JourneyRecoveryController.onPageLoad()).toFuture
+          }
       }
+  }
+
+  //noinspection ScalaStyle
```
Comment (darbyo) [inline app/controllers/CheckYourAnswersController.scala:218]: Could we remove this comment and fix any subsequent warnings

File: app/controllers/CheckYourAnswersController.scala [hunk 4]
```diff
@@ -151,81 +156,143 @@ class CheckYourAnswersController @Inject()(
...
+  def getRedirect(errors: List[ValidationError], period: Period): List[Call] = {
+    errors.flatMap {
+      case DataMissingError(AllSalesFromNiQuery) =>
+        logger.error(s"Data missing - no data provided for NI sales")
+        Some(routes.CountryOfConsumptionFromNiController.onPageLoad(CheckMode, period, Index(0)))
```
Comment (darbyo) [inline app/controllers/CheckYourAnswersController.scala:223]: Is is right that this is always Index(0)?

File: app/controllers/CheckYourAnswersController.scala [hunk 5]
```diff
@@ -151,81 +156,143 @@ class CheckYourAnswersController @Inject()(
...
+      case DataMissingError(NiSalesAtVatRateQuery(countryIndex, vatRateIndex)) =>
+        logger.error(s"Data missing - net value of sales at vat rate ${vatRateIndex.position} for country ${countryIndex.position}")
+        Some(routes.NetValueOfSalesFromNiController.onPageLoad(CheckMode, period, countryIndex, vatRateIndex))
+      case DataMissingError(VatOnSalesFromNiQuery(countryIndex, vatRateIndex)) =>
+        logger.error(s"Data missing - vat charged on sales at vat rate ${vatRateIndex.position} for country ${countryIndex.position}")
+        Some(routes.VatOnSalesFromNiController.onPageLoad(CheckMode, period, countryIndex, vatRateIndex))
+
+      case DataMissingError(AllSalesFromEuQuery) =>
+        logger.error(s"Data missing - no data provided for EU sales")
+        Some(routes.CountryOfSaleFromEuController.onPageLoad(CheckMode, period, Index(0)))
```
Comment (darbyo) [inline app/controllers/CheckYourAnswersController.scala:236]: as per previous comment

File: app/controllers/CheckYourAnswersController.scala [hunk 6]
```diff
@@ -151,81 +156,143 @@ class CheckYourAnswersController @Inject()(
...
+      case DataMissingError(AllSalesToEuQuery(countryFromIndex)) =>
+        logger.error(s"Data missing - country of consumption from country ${countryFromIndex.position}")
+        Some(routes.CountryOfConsumptionFromEuController.onPageLoad(CheckMode, period, countryFromIndex, Index(0)))
```
Comment (darbyo) [inline app/controllers/CheckYourAnswersController.scala:239]: as per previous comment

File: app/controllers/CheckYourAnswersController.scala [hunk 7]
```diff
@@ -151,81 +156,143 @@ class CheckYourAnswersController @Inject()(
...
+          s"${countryToIndex.position} at vat rate ${vatRateIndex.position} ")
+        Some(routes.NetValueOfSalesFromEuController.onPageLoad(CheckMode, period, countryFromIndex, countryToIndex, vatRateIndex))
+      case DataMissingError(VatOnSalesFromEuQuery(countryFromIndex, countryToIndex, vatRateIndex)) =>
+        logger.error(s"Data missing - vat charged on sales from country ${countryFromIndex.position} to country " +
+          s"${countryToIndex.position} at vat rate ${vatRateIndex.position} ")
+        Some(routes.VatOnSalesFromEuController.onPageLoad(CheckMode, period, countryFromIndex, countryToIndex, vatRateIndex))
 
+      case DataMissingError(AllCorrectionPeriodsQuery) =>
+        logger.error(s"Data missing - no data provided for corrections")
+        Some(correctionsRoutes.CorrectionReturnPeriodController.onPageLoad(CheckMode, period, Index(0)))
```
Comment (darbyo) [inline app/controllers/CheckYourAnswersController.scala:254]: As per previous comment

File: app/controllers/CheckYourAnswersController.scala [hunk 8]
```diff
@@ -151,81 +156,143 @@ class CheckYourAnswersController @Inject()(
...
+      case DataMissingError(AllCorrectionCountriesQuery(periodIndex)) =>
+        logger.error(s"Data missing - no countries found for corrections to period ${periodIndex.position}")
+        Some(correctionsRoutes.CorrectionCountryController.onPageLoad(CheckMode, period, periodIndex, Index(0)))
```
Comment (darbyo) [inline app/controllers/CheckYourAnswersController.scala:257]: As per previous comment

File: app/controllers/CheckYourAnswersController.scala [hunk 9]
```diff
@@ -151,81 +156,143 @@ class CheckYourAnswersController @Inject()(
...
+
+  def submitReturn(vatReturnRequest: VatReturnRequest, correctionRequest: Option[CorrectionRequest], period: Period)
+                  (implicit request: DataRequest[AnyContent]): Future[Result] = {
+
+    val submission = correctionRequest match {
+      case Some(cr) => vatReturnConnector.submit(VatReturnWithCorrectionRequest(vatReturnRequest, cr))
+      case _ => vatReturnConnector.submit(vatReturnRequest)
+    }
+
+    submission.flatMap {
```
Comment (darbyo) [inline app/controllers/CheckYourAnswersController.scala:279]: The first 2 cases have the same outcome, could refactor the order to make it so we just have one? Also the tabbing is a bit out on this, makes it a little harder to read

File: app/controllers/SalesFromEuListController.scala [hunk 10]
```diff
@@ -48,25 +49,46 @@ class SalesFromEuListController @Inject()(
...
     implicit request =>
-      getNumberOfSalesFromEu {
-        number =>
+      withCompleteData[SalesFromEuWithOptionalVat](
+        data = getIncompleteFromEuSales _,
+        onFailure = (_: Seq[SalesFromEuWithOptionalVat]) => {
+        if(incompletePromptShown) {
+          firstIndexedIncompleteSaleFromEu match {
+            case Some(incompleteSalesFromCountry) if incompleteSalesFromCountry._1.salesFromCountry.isEmpty =>
+              Redirect(routes.CountryOfConsumptionFromEuController.onPageLoad( mode,  period, Index(incompleteSalesFromCountry._2), Index(0)))
```
Comment (darbyo) [inline app/controllers/SalesFromEuListController.scala:68]: Will it always be index(0)?

File: app/models/SalesFromEuWithOptionalVat.scala [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package models
+
+import play.api.libs.json.{Json, OFormat}
+
+// TODO: This will eventually be fleshed out to represent all the data of a sale from NI
```
Comment (darbyo) [inline app/models/SalesFromEuWithOptionalVat.scala:21]: Is this TODO outstanding?

File: app/models/VatRateAndSales.scala [hunk 12]
```diff
@@ -27,7 +27,14 @@ case class VatRateAndSales(
                     validFrom: LocalDate,
                     validUntil: Option[LocalDate] = None,
                     sales: Option[SalesAtVatRate]
-                  )
+                  ) {
+
+  lazy val rateForDisplay: String = if(rate.isWhole) {
+    rate.toString.split('.').headOption.getOrElse(rate.toString) + "%"
```
Comment (darbyo) [inline app/models/VatRateAndSales.scala:33]: Where is this used? Cant seem to find any usages. If its not can we remove? If it is, why wouldn't we show the full rate?

File: app/models/VatRateAndSalesWithOptionalVat.scala [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+case class VatRateAndSalesWithOptionalVat(
+                    rate: BigDecimal,
+                    rateType: VatRateType,
+                    validFrom: LocalDate,
+                    validUntil: Option[LocalDate] = None,
+                    sales: Option[SalesAtVatRateWithOptionalVat]
+                  ) {
+
+  lazy val rateForDisplay: String = if(rate.isWhole) {
+    rate.toString.split('.').headOption.getOrElse(rate.toString) + "%"
```
Comment (darbyo) [inline app/models/VatRateAndSalesWithOptionalVat.scala:33]: As per previous comment.

File: app/views/corrections/VatPeriodCorrectionsListView.scala.html [hunk 14]
```diff
@@ -37,15 +37,34 @@
...
+            </strong>
+        </div>
+        }
 
-            @addToList(periodList, itemSize = Long, "vatPeriodCorrectionsList.change.hidden", "vatPeriodCorrectionsList.remove.hidden")
+        @addToList(periodList, itemSize = Long, "vatPeriodCorrectionsList.change.hidden", "vatPeriodCorrectionsList.remove.hidden")
 
-            @button("site.continue",  period, controllers.corrections.routes.VatPeriodCorrectionsListController.onPageLoad(mode, period).url)
+        @if(incompletePeriods.isEmpty) {
+            @formHelper(action = controllers.corrections.routes.VatPeriodCorrectionsListController.onSubmit(mode, period, false), 'autoComplete -> "off") {
```
Comment (darbyo) [inline app/views/corrections/VatPeriodCorrectionsListView.scala.html:57]: I have noticed the setting of 'autoComplete' -> "off" in a lot of places I assume this is intentional, but what is the reason behind it?

**Review by darbyo — CHANGES_REQUESTED** [review]

File: app/connectors/VatReturnConnector.scala (hunk 1 above)
Comment (owen-opencast) [reply app/connectors/VatReturnConnector.scala:41]: It was part of the refactoring of the CYA page - the intention was to overload the submit method, it looks like this wasn't needed in the end! Would you like it changed back?

File: app/controllers/CheckYourAnswersController.scala (hunk 2 above)
Comment (owen-opencast) [reply app/controllers/CheckYourAnswersController.scala:184]: Done

File: app/controllers/CheckYourAnswersController.scala (hunk 4 above)
Comment (owen-opencast) [reply app/controllers/CheckYourAnswersController.scala:223]: Yes, this is for when a user has answered "Yes" to NI sales, but have provided no further data.

File: app/controllers/CheckYourAnswersController.scala (hunk 3 above)
Comment (owen-opencast) [reply app/controllers/CheckYourAnswersController.scala:218]: It was a cyclomatic complexity warning. As there are so many case clauses I can't see a simple way of fixing this warning without complicating things. I'm open to suggestions!

File: app/controllers/CheckYourAnswersController.scala (hunk 5 above)
Comment (owen-opencast) [reply app/controllers/CheckYourAnswersController.scala:236]: As above, but for EU sales

File: app/controllers/SalesFromEuListController.scala (hunk 10 above)
Comment (owen-opencast) [reply app/controllers/SalesFromEuListController.scala:68]: Yes, this is for when a country of sale has been specified, but no countries of consumption have been specified, therefore country of consumption index will always be 0

File: app/models/SalesFromEuWithOptionalVat.scala (hunk 11 above)
Comment (owen-opencast) [reply app/models/SalesFromEuWithOptionalVat.scala:21]: It is in a sense - this class is a duplicate of SalesFromEu, but with salesFromCountry changed to Optional. The TODO was added by "Wilson" in August - should we remove both?

File: app/models/VatRateAndSalesWithOptionalVat.scala (hunk 13 above)
Comment (owen-opencast) [reply app/models/VatRateAndSalesWithOptionalVat.scala:33]: As above

File: app/connectors/VatReturnConnector.scala (hunk 1 above)
Comment (darbyo) [reply app/connectors/VatReturnConnector.scala:41]: I think so, if the overloading isnt needed, lets have it more descriptive

File: app/models/SalesFromEuWithOptionalVat.scala (hunk 11 above)
Comment (darbyo) [reply app/models/SalesFromEuWithOptionalVat.scala:21]: I think we can remove both then

File: app/models/VatRateAndSales.scala (hunk 12 above)
Comment (darbyo) [reply app/models/VatRateAndSales.scala:33]: ah sorry, I didnt spot the check if its whole!

**Review by darbyo — APPROVED** [review]

**Pull request merged**